### ISSUE TITLE
Add rule to validate value types

### ISSUE THREAD

**Issue by mixmastamyk** [stylelint/stylelint#2777]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

feature request

> Which version of stylelint are you using?

8.0.0

> How are you running stylelint: 
CLI 

e.g. "CLI with `stylelint --config myconfig *.css`"

Yes, with rc.yaml file

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?
No

Hi, my media query (responsive css) wasn't working, couldn't figure out why, wasted almost an hour when finally figured out the number I pasted into it didn't have units.  Grrr.  This is what brought me to csslint, then stylelint.  

Can't seem to find a rule that requires them.  Might be a bit complicated, needs know where units are needed, and to work  with the "no units for 0" rule as well.  Media queries might be a good place to start, since units are needed and in obvious locations.

**Comment by ntwb** [comment]
Thanks for creating this issue @mixmastamyk but could you supply some example code and explain the behaviour of what results you currently get versus what you expect to get.

(There are sections for adding these in the new issue template)

**Comment by mixmastamyk** [comment]
Yes, had css where I added a new value to a media query:

```css 
@media screen and (min-width: 768) {
    body {
        background-color: dodgerblue;  /* example */
    }
}
```

For what seemed like an hour I poked around at this and that, it just didn't work no matter what I did.  

Finally I went to a website cut and pasted the syntax again and studied the two intensly until I found the discrepancy:  the media query won't work at all until you add a unit, say 768**px** to the end of the value.  The browser I was using didn't feel obliged to say anything, though perhaps one of them does?

Anyway I'd love a linter rule to catch invalid values such as these that require units.  Thanks!

**Comment by evilebottnawi** [comment]
Perhaps concentration on validation rules will give a fresh breath of air for stylelint (prettier now solves many problems with newlines and spaces). 
@stylelint/core what do your think about future of stylelint and rules?

**Comment by lukaw3d** [comment]
+1; In my example this passed linter
```css
element {
    width: 20;
    height: 20px;
}
```

**Comment by jeddy3** [comment]
> @stylelint/core what do your think about future of stylelint and rules?

As with https://github.com/stylelint/stylelint/issues/2816, in the [VISON.md](https://github.com/stylelint/stylelint/blob/master/VISION.md) we have:

> Invalid code is best handled by emerging dedicated tools e.g. csstree - a language parser with syntax validation. As a stop-gap, while these tools mature, provide invalid code rules for the simplest of cases.

I'm not sure this is the "simplest of cases". Perhaps it's better to add this type of validation to a tool like `csstree` (or `css-values`) as, I believe, we're suggesting [validating value types](https://www.w3.org/TR/css3-values/) and property/value pairs - which looks complex.

Do we wish to continue with the approach outlined in the VISION, or do we need to revise it? I think helping to push a tool like csstree forward, and then making it easy to use within stylelint as a plugin still seems, to me, the most prudent approach.

**Comment by ntwb** [comment]
> Do we wish to continue with the approach outlined in the VISION, or do we need to revise it? I think helping to push a tool like csstree forward, and then making it easy to use within stylelint as a plugin still seems, to me, the most prudent approach.

+1 on this, otherwise we'd be re-inventing the wheel in this case

**Comment by jeddy3** [comment]
I'm going to close this for now, as keeping with the approach in the [VISION.md](https://github.com/stylelint/stylelint/blob/master/VISION.md) makes sense. Let's help push the syntax validation in csstree forward. We can promote the plugin in issue (and revise our documentation to emphasise it once it matures).

@mixmastamyk Keep an eye on the [stylelint-csstree-validator](https://github.com/csstree/stylelint-validator) plugin. It currently validates the property & value pairs _of declarations_, but I suspect it'll support property & value pairs for _media features_ at some point. 

@lukaw3d The [stylelint-csstree-validator](https://github.com/csstree/stylelint-validator) plugin will validate property & value pairs _of declarations_ and will warn for `width: 0` :)
